cli/tools/bundle/html.rs

```diff
@@ -0,0 +1,358 @@
+    BytePos(1),
+  );
+  let mut errors = Vec::new();
+  let mut doc = swc_html::parser::parse_file_as_document(
+    &file,
+    swc_html::parser::parser::ParserConfig {
+      ..Default::default()
+    },
+    &mut errors,
+  )
+  .unwrap();
+
+  let mut scripts = collect_scripts(&mut doc);
+
+  let mut temp_module = String::new();
+  for script in &mut scripts {
+    if let Some(src) = &mut script.src {
+      let src = src.trim_start_matches('/');
+      let path = path.parent().unwrap().join(src);
+
+      temp_module
+        .push_str(&format!("import \"{}\";\n", path.to_string_lossy()));
+      script.resolved_path = Some(path);
+    }
+  }
+
+  Ok(HtmlEntrypoint {
+    path: path.to_path_buf(),
+    scripts,
+    temp_module,
+    doc,
+  })
+}
+
+pub struct Remover {
+  to_inject: Script,
+  css_to_inject_path: Option<String>,
+
+  injected: bool,
+}
+
+impl VisitHtml for Remover {
+  fn visit_element(&mut self, e: &mut Element) {
+    if e.tag_name == "head" {
+      if !self.injected {
+        self.injected = true;
+        e.children.push(Child::Element(self.to_inject.to_element()));
+        if let Some(css_to_inject_path) = &self.css_to_inject_path {
+          e.children.push(Child::Element(Element {
+            attributes: vec![
+              make_attr("rel", Some("stylesheet")),
+              make_attr("crossorigin", None),
+              make_attr("href", Some(css_to_inject_path)),
+            ],
+            children: vec![],
+            content: None,
+            span: Span::default(),
+            tag_name: atom!("link"),
+            namespace: Namespace::HTML,
+            is_self_closing: true,
+          }));
+        }
+      }
+    }
+    let mut remove = Vec::new();
+    for (i, e) in &mut e.children.iter_mut().enumerate() {
+      match e {
+        Child::Element(element) => {
+          if element.tag_name == "script" {
+            if let Some(src) = get_attr(element, "src")
+              && Some(src) != self.to_inject.src.as_deref()
+            {
+              remove.push(i);
+            }
+          } else {
+            self.visit_element(element);
+          }
+        }
+        _ => {}
+      }
+    }
+
+    let mut new_children = Vec::new();
+    for (i, e) in &mut e.children.iter_mut().enumerate() {
+      if !remove.contains(&i) {
+        new_children.push(e.clone());
+      }
+    }
+    e.children = new_children;
+  }
+}
+
+impl HtmlEntrypoint {
+  pub fn to_html(&self) -> String {
+    let mut out = String::new();
+    let mut s = BasicHtmlWriter::new(
+      &mut out,
+      None,
+      swc_html::codegen::writer::basic::BasicHtmlWriterConfig {
+        ..Default::default()
+      },
+    );
+    let mut codegen = swc_html::codegen::CodeGenerator::new(
+      &mut s,
+      swc_html::codegen::CodegenConfig {
+        ..Default::default()
+      },
+    );
+    codegen.emit(&self.doc).unwrap();
+    out
+  }
+  pub fn patch_html_with_response<'a>(
```

Comment:
Super clever, we need to test this - I understand it's probably too cumbersome to unit test this, so having a few integration tests would be great. Thanks!